### ISSUE TITLE
Improve detection of java executable

### ISSUE THREAD

**Issue by angelozerr** [redhat-developer/vscode-xml#13]
In my case I have an installed java but not a JAVA_HOME. My java path is in my Windows PATH variable.

To use this java without configuring VSCode preferences, you could search it:

 * with `where java` command in OS Windows
 * with `which java` command for other OS.

This idea was done in Eclipse Bluesky to find the node process. Here the Java code:

```java
public static String getNodeJsLocation() {
		String res = "/path/to/node";
		String[] command = new String[] {"/bin/bash", "-c", "which node"};
		if (Platform.getOS().equals(Platform.OS_WIN32)) {
			command = new String[] {"cmd", "/c", "where node"};
		}
		BufferedReader reader = null;
		try {
			Process p = Runtime.getRuntime().exec(command);
			reader = new BufferedReader(new InputStreamReader(p.getInputStream()));
			res = reader.readLine();
		} catch (IOException e) {
			Activator.getDefault().getLog().log(
					new Status(IStatus.ERROR, Activator.getDefault().getBundle().getSymbolicName(), e.getMessage(), e));
		} finally {
			IOUtils.closeQuietly(reader);
		}

		// Try default install path as last resort
		if (res == null && Platform.getOS().equals(Platform.OS_MACOSX)) {
			res = "/usr/local/bin/node";
		}

		if (res != null && Files.exists(Paths.get(res))) {
			return res;
		} else if (!alreadyWarned){
			warnNodeJSMissing();
			alreadyWarned = true;
		}
		return null;
	}
```

It should be cool to adapt this Java code in TypeScript and to discover java instead of node.

**Comment by fbricon** [comment]
@angelozerr can you please try again with the latest build from http://download.jboss.org/jbosstools/vscode-xml/staging/?C=M;O=D. @snjeza checked this actually works on windows

**Comment by angelozerr** [comment]
I have installed  vscode-xml-0.1.0-206.vsix but it doesn't work. I must define my java.home in preferences again.

**Comment by fbricon** [comment]
where is your java directory located?
can you paste the content of %PATH%?

the java detection from the path is provided by https://github.com/redhat-developer/vscode-xml/blob/master/src/requirements.ts#L9 
Only way to debug it would be to sideload the vscode-xml extension, similar to [vscode-java](https://github.com/redhat-developer/vscode-java/blob/master/CONTRIBUTING.md#sideloading). (@NikolasKomonen will certainly add a contributing guide \*wink\* \*wink\*)

**Comment by NikolasKomonen** [comment]
> (@NikolasKomonen will certainly add a contributing guide *wink* *wink*)

He sure will

**Comment by angelozerr** [comment]
Thanks for the guide @NikolasKomonen 

Here my path:

![image](https://user-images.githubusercontent.com/1932211/45378333-72ab9180-b5fd-11e8-92e7-88e6f1fa88eb.png)

**Comment by snjeza** [comment]
> Have you some info to debug the typescript extension?

Windows
- cd <some_directory>
- clone lsp4xml
- clone vscode-xml
- cd lsp4xml
- mvn clean verify -DskipTests
- cd ../vscode-xml
- npm install
- npm run compile
- mkdir server
- cp ../lsp4xml/org.eclipse.lsp4xml/target/org.eclipse.lsp4xml-all.jar server
- code .
- debug 'Launch Extension'

Linux/Mac
- cd <some_directory>
- clone lsp4xml
- clone vscode-xml
- cd vscode-xml
- npm install
- npm run compile
- npm run build-server
- code .
- debug 'Launch Extension'

**Comment by snjeza** [comment]
@angelozerr Are you using a 32-bit JDK?

**Comment by angelozerr** [comment]
@snjeza see my PR https://github.com/redhat-developer/vscode-xml/pull/33 which must not be merged but which show you how I have fixed the problem with my Window context. I'm not sure if you are agree to support this javapath Oracle context. But JRE should be supported I think.

**Comment by snjeza** [comment]
I can't reproduce the issue.
@angelozerr  Your PR doesn't work if JAVA_HOME is set. It also doesn't work on Linux.
You can create an issue against https://github.com/jsdevel/node-find-java-home.

**Comment by fbricon** [comment]
The whole javapath thing seems to be specific to Oracle JRE on windows. 
For now, since we documented JDK as a requirement, I don't think this is a pressing issue if we want to release.

**Comment by aaroneg** [comment]
With the latest AdoptOpenJDK installed (OpenJDK14U-jdk_x64_windows_hotspot_14.0.1_7) this extension complains that jdk needs to be installed. I'm not sure if it's failing to detect or just popping up a warning. 
![image](https://user-images.githubusercontent.com/1396671/81217603-7c3bf700-8fa2-11ea-8f19-945abb92488d.png)

**Comment by fbricon** [comment]
What is the result of `java -version`?

**Comment by aaroneg** [comment]
> openjdk version "14.0.1" 2020-04-14
OpenJDK Runtime Environment AdoptOpenJDK (build 14.0.1+7)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 14.0.1+7, mixed mode, sharing)

**Comment by fbricon** [comment]
@aaroneg have you tried setting the "xml.java.home" preference?

**Comment by aaroneg** [comment]
No, that seems like it's out of scope for the end-user to do. The java detection should work properly instead.

**Comment by fbricon** [comment]
How did you install your jdk? Is JAVA_HOME set? Does it work if you launch vscode from a terminal?

**Comment by fbricon** [comment]
What is the result of `where java`?

**Comment by aaroneg** [comment]
1) https://adoptopenjdk.net/ just grabbing the version appropriate to the operating system
2) `JAVA_HOME=C:\Program Files\AdoptOpenJDK\jdk-14.0.1.7-hotspot\`
3) the extension is no longer installed because of this problem
4) 
```
C:\Users\aaroneg>where java
C:\Program Files\AdoptOpenJDK\jdk-14.0.1.7-hotspot\bin\java.exe
C:\ProgramData\Oracle\Java\javapath\java.exe
```

**Comment by bearslumber** [comment]
I get the same error as aaroneg.

I have tried Oracle Java, Zulu Java, and even Redhat Java.

I am currently switching to another plugin because I can no longer use this one.

**Comment by bearslumber** [comment]
![image](https://user-images.githubusercontent.com/22350074/82666888-b1954580-9c2e-11ea-99c4-c0840bb373e0.png)

**Comment by bearslumber** [comment]
![image](https://user-images.githubusercontent.com/22350074/82666927-ca056000-9c2e-11ea-88dc-880a6f0688a0.png)

![image](https://user-images.githubusercontent.com/22350074/82667092-0d5fce80-9c2f-11ea-9bce-ec8a4ace7a3f.png)



Having got the error that aaroneg is getting I set the xml.java.home variable in VS Code settings as suggested in the help files. 

Still no joy. Very frustrating. :(

**Comment by angelozerr** [comment]
> Any help is greatly appreciated.

Try to remove `\\jre\\bin` from your `xml.java.home`

**Comment by angelozerr** [comment]
It should be fixed
